### PR TITLE
Running builder (content generator) functions in subprocesses on Windows

### LINKED ISSUE #5042: On Windows at least building with a -j > 1 scons parameter fails with a file not found even though the file exists
**Operating system or device - Godot version:**

Windows 10 Pro 64-bit, Godot source 64-bit build (tested with master branch)

Visual Studio 2015 Community, Windows 10 SDK Kit 10586.

Python 2.7.11, Scons 2.5.0, and Pywin32 220 all these tools are 32-bit.

**Issue description** (what happened, and what was expected):

When I issue the command:

scons platform=windows target=release_debug -jx

Where "x" is a number > 1 then the build process will eventually give a file not found error.  Going to where it says the file is not found it is present.  Specifying -j1 on the build line works fine.

**Steps to reproduce:**

scons platform=windows target=release_debug -j2

### PR DESCRIPTION
**Pull request by viktor-ferenczi**
- Refactored all builder (make_*) functions into separate Python modules along the build tree
- Introduced utility function to wrap all invocations into subprocesses on Windows
- Introduced stub to use the builder modules as a stand alone scripts to execute a selected function

**Background**

There is a problem with file handles related to writing generated content (`*.gen.h` and `*.gen.cpp`)
on Windows, which randomly causes a `SHARING VIOLATION` error to the compiler resulting in flaky
builds. Running all such content generator in a new subprocess instead of directly inside the build 
script works around the issue.

**Q&A**

- Root cause?

The root cause of the issue is unknown, but it is specific to how file handles are somehow shared with child processes, even if the files are closed before spawning the child. Attempting to disable file handle sharing by setting the `HANDLE_FLAG_INHERIT` option to zero on all file handles involved did not work out.

- Why not using the multiprocessing module?

Yes, I tried the multiprocessing module. It did not work due to conflict with SCons on `cPickle`.
Suggested workaround did not fully work either.

- Why `run_in_subprocess` is not used as a decorator on the builder functions?

One reason is to make it explicit in the Builder definition that the function will run in a subprocess. 

Other reason is that the original builder function need to be called in the subprocess, which would need saving that function on the wrapper and special casing it there to call the original function while not running inside the main build script. It would involve adding magic auto detection code, also one more symbol to import in the builder modules. I decided not doing that, despite the decorator syntax would look nicer. I can still implement this per request if you would like it better that way.

- Why the builder functions are wrapped on non-Windows platforms (osx, x11)?

Using the `run_in_subprocess` wrapper on `osx` and `x11` platforms as well for consistency, but it 
does not actually create subprocesses there. In case of running a cross-compilation on 
Windows they would still be used, but likely it will not happen in practice. What counts is 
the platform the build itself is running on, not the target platform.

Fixes #5042

### CODE REVIEW

**Comment by akien-mga** [comment]
CC @garyo for review.

File: compat.py [hunk 1]
```diff
@@ -62,3 +70,60 @@ def escape_string(s):
                 result += chr(c)
         return result
 
+
+def run_in_subprocess(builder_function):
```
Comment (akien-mga) [inline compat.py]: I think the `compat` module was meant only for Python 2/3 compatibility definitions. This should maybe go in `methods`?

Comment (viktor-ferenczi) [reply compat.py]: This is more of platform compatibility. Maybe a new module would work best. Something like `plat_compat.py` or so.

Comment (akien-mga) [reply compat.py]: I don't know, `methods.py` already has some platform-specific methods. Maybe you could make a new `platform_methods.py` module and move the relevant `methods.py` methods there too.

Comment (viktor-ferenczi) [reply compat.py]: Good idea, but that would be unrelated to the scope of this PR and the ticket behind. 

To make it clear I suggest adding `platform_methods.py` with the new functions introduced here. Then we can move the rest of functions between the modules in a separate PR.

Comment (akien-mga) [reply compat.py]: Fine by me :)

**Comment by viktor-ferenczi** [comment]
Notes on resolving conflicts with two previously merged PRs:
- Introduction of the authors, donors and license generated header files. 
- [Introduction of env.CommandNoCache](https://github.com/godotengine/godot/pull/19695), which [fixed a Travis cache corruption problem](https://github.com/godotengine/godot/issues/19665). As far as I understand this change is unrelated to the issue fixed in this PR. Reason is that the Travis cache corruption issue affected non-Windows builds.

**Comment by viktor-ferenczi** [comment]
The last Travis build were affected by the "cache full" issue, but the NoCache change I merged did not solve it:

```
Current cache limit is 1.0 GB (used: 1.0 GB)
Purging 5 files from cache...
```

There seem to be no way for me to restart that specific build with clean cache.

Any ideas?

**Comment by akien-mga** [comment]
> `c:\users\...\development\godot\drivers\gles2\rasterizer_scene_gles2.h(36): fatal error C1083: Cannot open include file: 'shaders/scene.glsl.gen.h': No such file or directory`

Actually, this might not invalidate this patch, it's just that it needs to be completed to also handle the generation of the shader headers in `drivers/gles2` and `drivers/gles3`. Could you do that @viktor-ferenczi?

**Comment by akien-mga** [comment]
And basically all `.gen.*` files need to get the same treatment I guess.

**Comment by viktor-ferenczi** [comment]
Rebased to latest master. Working on enrolling all newly added `*.gen.*` files to the subprocess based build.

**Comment by mhilbrunner** [comment]
I can confirm that this seems to fix the issue for me.

I tested this on two different Windows machines where this fails on `master`, and it does fine even on completely clean builds using 8 or 16 cores.

**Comment by mhilbrunner** [comment]
This has a conflict that needs to be fixed, then I'd be strongly in favor of merging this one.

**Comment by viktor-ferenczi** [comment]
This PR is moving most of the generator code from `SConstruct / SCsub / methods.py` to corresponding `*_builders.py` modules.

Could somebody else please verify that no code were missed in the process?

**Comment by akien-mga** [comment]
Awesome work, thanks a ton @viktor-ferenczi!

**Comment by viktor-ferenczi** [comment]
Thank you. It was a pleasure to help your awesome project.

**Comment by moiman100** [comment]
Trying to build master after this commit causes this https://puu.sh/B4B6M/2243083766.txt
Windows 10 64-bit

**Comment by bdbaddog** [comment]
Rather than moving your python generator functions such that they can be run separately and then running them via a builder which calls subprocess, wouldn't it be simpler to just have your builders call those python scripts via normal builders?  Since the issue (if I understand correctly) was that you had blobs of python logic in your builders in the same python process and GIL issues, and you've resolve that by making the generators independent scripts.

So your Action would now be "python <generator_script.py> <arguments>"?  This would run the python scripts in a separate process

**Comment by viktor-ferenczi** [comment]
@moiman100 Problem is that the build is not tested by CI on Python 3.7 and I did not do it either (my bad). 

Type basestring is not available on Python 3.7, so the issue. I'm going to add a new PR to fix Python 3.5+ compatiblity.

**Comment by viktor-ferenczi** [comment]
Perfect, thanks for the update and sorry for the mistake.

**Comment by marcelofg55** [comment]
I'm getting a build error on OS X when building with `scons -j4 platform=osx bits=64 debug_symbols=full separate_debug_symbols=yes dev=yes`:

```
[ 99%] progress_finish(["progress_finish"], [])
[100%] make_debug_osx(["bin/godot.osx.tools.64"], ["platform/osx/crash_handler_osx.osx.tools.64.o", "platform/osx/os_osx.osx.tools.64.o", "platform/osx/godot_main_osx.osx.tools.64.o", "platform/osx/sem_osx.osx.tools.64.o", "platform/osx/dir_access_osx.osx.tools.64.o", "platform/osx/joypad_osx.osx.tools.64.o", "platform/osx/power_osx.osx.tools.64.o"])
scons: *** [bin/godot.osx.tools.64] TypeError : 'NoneType' object has no attribute '__getitem__'
Traceback (most recent call last):
  File "/usr/local/Cellar/scons/3.0.0/libexec/scons-local/SCons/Action.py", line 1197, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/Volumes/Data/git/godot_3.0/platform_methods.py", line 22, in wrapper
    return builder_function(target, source, None)
  File "/Volumes/Data/git/godot_3.0/platform/osx/platform_osx_builders.py", line 11, in make_debug_osx
    if (env["macports_clang"] != 'no'):
TypeError: 'NoneType' object has no attribute '__getitem__'
scons: building terminated because of errors.
ERROR!
```

**Comment by viktor-ferenczi** [comment]
It is because the `env` parameter is not passed over to the subprocesses. It is because there were no usages in Windows nor Linux builds. It looks like I missed the reference on Mac.

Possible solutions:
1. Pass `env` as a dictionary (converting if needed)
2. Running the builds inside the process (instead of in subprocess) on non-Windows platforms

I think 2. would be simpler. It needs a simple platform condition to the `run_in_subprocess` method to return the builder function as is if not on Windows.

**Comment by viktor-ferenczi** [comment]
Fix is in PR #20617

**Comment by viktor-ferenczi** [comment]
PR #20617 has passed all checks, tested on Mac by @marcelofg55 
It is ready for code review.

**Comment by viktor-ferenczi** [comment]
PR #20617 completed, so the Mac build error caused by this PR has been fixed.

**Pull request merged**